src/VisualStudio/Core/Def/Implementation/DesignerAttribute/DesignerAttributeIncrementalAnalyzer.cs

```diff
@@ -53,198 +50,190 @@ internal partial class DesignerAttributeIncrementalAnalyzer : ForegroundThreadAf
+                   document.Project.CanReusePersistedDependentProjectVersion(semanticVersion, existingData.SemanticVersion);
+        }
 
-            foreach (var newKvp in newFileToResult)
+        private async Task RegisterDesignerAttributeAndSaveStateAsync(
+            Document document, VersionStamp textVersion, VersionStamp semanticVersion, Optional<string> designerAttributeArgumentOpt, CancellationToken cancellationToken)
+        {
+            if (!designerAttributeArgumentOpt.HasValue)
             {
-                // 1) If this result is for a new document.  We always need to report it
-                // 2) If both the old and new data have this result, then report it if it is different.
-                var filePath = newKvp.Key;
-                var newResult = newKvp.Value;
-
-                if (!oldFileToResult.TryGetValue(filePath, out var oldResult) ||
-                    !newResult.Equals(oldResult))
-                {
-                    difference.Add(filePath, newResult);
-                }
+                // no value means it couldn't determine whether this document has designer attribute or not.
+                // one of such case is when base type is error type.
+                return;
             }
 
-            return difference.ToImmutable();
+            var data = new Data(textVersion, semanticVersion, designerAttributeArgumentOpt.Value);
+            await _state.PersistAsync(document, data, cancellationToken).ConfigureAwait(false);
+
+            RegisterDesignerAttribute(document, designerAttributeArgumentOpt.Value);
         }
 
         private void RegisterDesignerAttribute(Document document, string designerAttributeArgument)
         {
-            var workspace = (VisualStudioWorkspaceImpl)document.Project.Solution.Workspace;
-
-            var vsDocument = workspace.GetHostDocument(document.Id);
-            if (vsDocument == null)
+            if (!_state.Update(document.Id, designerAttributeArgument))
             {
+                // value is not updated, meaning we are trying to report same value as before.
+                // we don't need to do anything. 
+                //
+                // I kept this since existing code had this, but since we check what platform has before
+                // reporting it, this might be redundant.
                 return;
             }
 
-            uint itemId = vsDocument.GetItemId();
-            if (itemId == (uint)VSConstants.VSITEMID.Nil)
+            var workspace = document.Project.Solution.Workspace as VisualStudioWorkspaceImpl;
+            if (workspace == null)
             {
-                // it is no longer part of the solution
                 return;
             }
 
-            if (ErrorHandler.Succeeded(vsDocument.Project.Hierarchy.GetProperty(itemId, (int)__VSHPROPID.VSHPROPID_ItemSubType, out var currentValue)))
+            var documentId = document.Id;
+            _notificationService.RegisterNotification(() =>
```

Comment:
:bulb: We should consider submitting items where `designerAttributeArgument` is not null or empty at higher priority than the rest, to minimize the delay from solution load to when designers can be used.